File: njli/platform/cmake.in/ldoc.in/PhysicsBody.lua

----
-- @file PhysicsBody


---- Brief description.
-- @return <#return value description#>
function PhysicsBody:PhysicsBody()
end

---- Brief description.
-- @return <#return value description#>
function PhysicsBody:PhysicsBody()
end

---- Brief description.
-- @return <#return value description#>
function PhysicsBody:PhysicsBody()
end

---- Brief description.
-- @return <#return value description#>
function PhysicsBody:BT_DECLARE_ALIGNED_ALLOCATOR()
end

---- Brief description.
-- @return <#return value description#>
function PhysicsBody:operator=()
end

---- Brief description.
-- <#Description#>
-- @return <#return value description#>
function PhysicsBody:calculateSerializeBufferSize()
end

---- Brief description.
-- @author James Folk, 16-02-11 15:02:51
-- <#Description#>
-- @param dataBuffer <#dataBuffer description#>
-- @param btSerializer <#btSerializer description#>
-- @return 
function PhysicsBody:serialize(dataBuffer, btSerializer)
end

---- Brief description.
-- <#Description#>
-- @return <#return value description#>
function PhysicsBody:getClassName()
end

---- Brief description.
-- <#Description#>
-- @return <#return value description#>
function PhysicsBody:getType()
end

---- Brief description.
-- <#Description#>
-- @param body <#body description#>
-- @param btManifoldPoint&pt <#btManifoldPoint&pt description#>
-- @return <#return value description#>
function PhysicsBody:handleCollide(body, btManifoldPoint&pt)
end

---- Brief description.
-- <#Description#>
-- @param body <#body description#>
-- @param dispatchInfo <#dispatchInfo description#>
-- @return <#return value description#>
function PhysicsBody:handleCollisionNear(body, dispatchInfo)
end

---- Brief description.
-- <#Description#>
-- @param shape <#shape description#>
-- @return <#return value description#>
function PhysicsBody:addPhysicsShape(shape)
end

---- Brief description.
-- <#Description#>
-- @return <#return value description#>
function PhysicsBody:removePhysicsShape()
end

---- Brief description.
-- @return <#return value description#>
function PhysicsBody:getPhysicsShape()
end

---- Brief description.
-- <#Description#>
-- @return <#return value description#>
function PhysicsBody:getPhysicsShape()
end

---- Brief description.
-- <#Description#>
-- @param transform <#transform description#>
-- @return <#return value description#>
function PhysicsBody:setWorldTransform(transform)
end

---- Brief description.
-- <#Description#>
-- @return <#return value description#>
function PhysicsBody:getWorldTransform()
end

---- Brief description.
-- <#Description#>
-- @param friction <#friction description#>
-- @return <#return value description#>
function PhysicsBody:setFriction(friction)
end

---- Brief description.
-- <#Description#>
-- @return <#return value description#>
function PhysicsBody:getFriction()
end

---- Brief description.
-- <#Description#>
-- @param friction <#friction description#>
-- @return <#return value description#>
function PhysicsBody:setAngularFriction(friction)
end

---- Brief description.
-- <#Description#>
-- @return <#return value description#>
function PhysicsBody:getAngularFriction()
end

---- Brief description.
-- <#Description#>
-- @param restitution <#restitution description#>
-- @return <#return value description#>
function PhysicsBody:setRestitution(restitution)
end

---- Brief description.
-- <#Description#>
-- @return <#return value description#>
function PhysicsBody:getRestitution()
end

---- Brief description.
-- <#Description#>
-- @param type <
-- @return <#return value description#>
function PhysicsBody:setPhysicsType(type)
end

---- Brief description.
-- <#Description#>
-- @return <#return value description#>
function PhysicsBody:getPhysicsType()
end

---- Brief description.
-- <#Description#>
-- @param categoryBitMask <#categoryBitMask description#>
-- @return <#return value description#>
function PhysicsBody:setCollisionGroup(categoryBitMask)
end

---- Brief description.
-- <#Description#>
-- @return <#return value description#>
function PhysicsBody:getCollisionGroup()
end

---- Brief description.
-- <#Description#>
-- @param categoryBitMask <#categoryBitMask description#>
-- @return <#return value description#>
function PhysicsBody:setCollisionMask(categoryBitMask)
end

---- Brief description.
-- <#Description#>
-- @return <#return value description#>
function PhysicsBody:getCollisionMask()
end

---- Brief description.
-- <#Description#>
-- @return <#return value description#>
function PhysicsBody:setStaticPhysics()
end

---- Brief description.
-- <#Description#>
-- @return <#return value description#>
function PhysicsBody:setKinematicPhysics()
end

---- Brief description.
-- <#Description#>
-- @return <#return value description#>
function PhysicsBody:setDynamicPhysics()
end

---- Brief description.
-- <#Description#>
-- @param enable <#enable description#>
-- @return <#return value description#>
function PhysicsBody:enableContactResponse(enable)
end

---- Brief description.
-- <#Description#>
-- @param enable <#enable description#>
-- @return <#return value description#>
function PhysicsBody:enableHandleCollideCallback(enable)
end

---- Brief description.
-- <#Description#>
-- @return <#return value description#>
function PhysicsBody:isStaticPhysics()
end

---- Brief description.
-- <#Description#>
-- @return <#return value description#>
function PhysicsBody:isKinematicPhysics()
end

---- Brief description.
-- <#Description#>
-- @return <#return value description#>
function PhysicsBody:isDynamicPhysics()
end

---- Brief description.
-- <#Description#>
-- @return <#return value description#>
function PhysicsBody:isResting()
end

---- Brief description.
-- <#Description#>
-- @param enable <#enable description#>
-- @return <#return value description#>
function PhysicsBody:enableResting(enable)
end

---- Brief description.
-- <#Description#>
-- @return <#return value description#>
function PhysicsBody:getVelocity()
end

---- Brief description.
-- <#Description#>
-- @return <#return value description#>
function PhysicsBody:getAngularVelocity()
end

---- Brief description.
-- <#Description#>
-- @return <#return value description#>
function PhysicsBody:isInWorld()
end

---- Brief description.
-- <#Description#>
-- @param physicsConstraints <#physicsConstraints description#>
-- @return <#return value description#>
function PhysicsBody:getPhysicsConstraints(physicsConstraints)
end

---- Brief description.
-- <#Description#>
-- @param object <#object description#>
-- @return <#return value description#>
function PhysicsBody:getPhysicsConstraintIndex(object)
end

---- Brief description.
-- <#Description#>
-- @param index <#index description#>
-- @return <#return value description#>
function PhysicsBody:getPhysicsConstraint(index)
end

---- Brief description.
-- <#Description#>
-- @param index <#index description#>
-- @return <#return value description#>
function PhysicsBody:getPhysicsConstraint(index)
end

---- Brief description.
-- @return <#return value description#>
function PhysicsBody:hasPhysicsConstraint()
end

---- Brief description.
-- <#Description#>
-- @return <#return value description#>
function PhysicsBody:getParent()
end

---- Brief description.
-- <#Description#>
-- @return <#return value description#>
function PhysicsBody:getParent()
end

---- Brief description.
-- <#Description#>
-- @return <#return value description#>
function PhysicsBody:getCollisionFlags()
end

---- Brief description.
-- <#Description#>
-- @return <#return value description#>
function PhysicsBody:getActivationState()
end

---- Brief description.
-- <#Description#>
-- @param type <
-- @return <#return value description#>
function NJLI.PhysicsBody.create(type)
end

---- Brief description.
-- <#Description#>
-- @param object <#object description#>
-- @return <#return value description#>
function NJLI.PhysicsBody.destroy(object)
end

---- Brief description.
-- <#Description#>
-- @param object <#object description#>
-- @param L <#L description#>
-- @param stack_index <#stack_index description#>
-- @return <#return value description#>
function NJLI.PhysicsBody.load(object, L, stack_index)
end

---- Brief description.
-- <#Description#>
-- @return <#return value description#>
function NJLI.PhysicsBody.type()
end

---- Brief description.
-- @return <#return value description#>
function NJLI.PhysicsBody.getCollisionObject()
end

---- Brief description.
-- @return <#return value description#>
function NJLI.PhysicsBody.getCollisionObject()
end

---- Brief description.
-- @return <#return value description#>
function NJLI.PhysicsBody.setCollisionObject()
end

---- Brief description.
-- @return <#return value description#>
function NJLI.PhysicsBody.addPhysicsBody()
end

---- Brief description.
-- @return <#return value description#>
function NJLI.PhysicsBody.removePhysicsBody()
end

---- Brief description.
-- <#Description#>
-- !!TODO: !!TODO: scene->addActivePhysicsConstraint(constraint);
-- @param constraint <#constraint description#>
-- @return <#return value description#>
function NJLI.PhysicsBody.addPhysicsConstraint(constraint)
end

---- Brief description.
-- <#Description#>
-- !!TODO: scene->removeActivePhysicsConstraint(*iter);
-- @param constraint <#constraint description#>
-- @return <#return value description#>
function NJLI.PhysicsBody.removePhysicsConstraint(constraint)
end

---- Brief description.
-- <#Description#>
-- <#Description#>
-- @return <#return value description#>
function NJLI.PhysicsBody.removeAllPhysicsConstraints()
end

File: njli/physics/body/PhysicsBody.cpp
//
//  PhysicsBody.cpp
//  JLIGameEngineTest
//
//  Created by James Folk on 11/22/14.
//  Copyright (c) 2014 James Folk. All rights reserved.
//

#include "PhysicsBody.h"
#include "JLIFactoryTypes.h"
#include "PhysicsShape.h"
#include "PhysicsWorld.h"
#include "World.h"
#include "btCollisionObject.h"
#include "btCollisionShape.h"
#include "btDispatcher.h"
#include "btManifoldPoint.h"
#include "btVector3.h"
#define FORMATSTRING "{\"njli::PhysicsBody\":[]}";

#include "PhysicsConstraint.h"

namespace njli
{
  PhysicsBody::PhysicsBody()
      : AbstractFactoryObject(this),
        //    m_PhysicsWorld(NULL),
        m_PhysicsShape(NULL),
        //    m_VelocityFactor(new btVector3()),
        //    m_AngularVelocityFactor(new btVector3()),
        //    m_njliPhysicsType(JLI_PHYSICS_TYPE_Dynamic),
        //    m_Mass(1.0f),
        //    m_Charge(0.0f),
        //    m_Friction(1.0f),
        //    m_AngularFriction(1.0f),
        //    m_Restitution(1.0f),
        //    m_Damping(1.0f),
        //    m_AngularDamping(1.0f),
        m_CategoryBitMask(JLI_BIT_CATEGORY_ALL),
        m_CollisionBitMask(JLI_BIT_CATEGORY_ALL),
        //    m_TotalForce(new btVector3()),
        //    m_TotalAngularForce(new btVector3()),
        m_RestingAllowed(true), m_IsResting(false),
        m_CollisionFlags(btCollisionObject::CF_STATIC_OBJECT),
        m_ActivationState(ACTIVE_TAG)
  {
  }

  PhysicsBody::PhysicsBody(const AbstractBuilder &builder)
      : AbstractFactoryObject(this),
        //    m_PhysicsWorld(NULL),
        m_PhysicsShape(NULL),
        //    m_VelocityFactor(new btVector3()),
        //    m_AngularVelocityFactor(new btVector3()),
        //    m_njliPhysicsType(JLI_PHYSICS_TYPE_Dynamic),
        //    m_Mass(1.0f),
        //    m_Charge(0.0f),
        //    m_Friction(1.0f),
        //    m_AngularFriction(1.0f),
        //    m_Restitution(1.0f),
        //    m_Damping(1.0f),
        //    m_AngularDamping(1.0f),
        m_CategoryBitMask(JLI_BIT_CATEGORY_ALL),
        m_CollisionBitMask(JLI_BIT_CATEGORY_ALL),
        //    m_TotalForce(new btVector3()),
        //    m_TotalAngularForce(new btVector3()),
        m_RestingAllowed(true), m_IsResting(false),
        m_CollisionFlags(btCollisionObject::CF_STATIC_OBJECT),
        m_ActivationState(ACTIVE_TAG)
  {
  }

  PhysicsBody::PhysicsBody(const PhysicsBody &copy)
      : AbstractFactoryObject(this),
        //    m_PhysicsWorld(NULL),
        m_PhysicsShape(NULL),
        //    m_VelocityFactor(new btVector3()),
        //    m_AngularVelocityFactor(new btVector3()),
        //    m_njliPhysicsType(JLI_PHYSICS_TYPE_Dynamic),
        //    m_Mass(1.0f),
        //    m_Charge(0.0f),
        //    m_Friction(1.0f),
        //    m_AngularFriction(1.0f),
        //    m_Restitution(1.0f),
        //    m_Damping(1.0f),
        //    m_AngularDamping(1.0f),
        m_CategoryBitMask(JLI_BIT_CATEGORY_ALL),
        m_CollisionBitMask(JLI_BIT_CATEGORY_ALL),
        //    m_TotalForce(new btVector3()),
        //    m_TotalAngularForce(new btVector3()),
        m_RestingAllowed(true), m_IsResting(false),
        m_CollisionFlags(btCollisionObject::CF_STATIC_OBJECT),
        m_ActivationState(ACTIVE_TAG)
  {
  }

  PhysicsBody::~PhysicsBody() {}

  PhysicsBody &PhysicsBody::operator=(const PhysicsBody &rhs)
  {
    if (this != &rhs)
      {
        // TODO:implement....
      }
    return *this;
  }

  PhysicsBody *PhysicsBody::create(u32 type)
  {
    SDL_assert(type == JLI_OBJECT_TYPE_PhysicsBodyRigid ||
               type == JLI_OBJECT_TYPE_PhysicsBodySoft ||
               type == JLI_OBJECT_TYPE_PhysicsBodyGhost);

    return dynamic_cast<PhysicsBody *>(
        World::getInstance()->getWorldFactory()->create(type));
  }

  void PhysicsBody::destroy(PhysicsBody *object)
  {
    if (object)
      {
        object->removePhysicsShape();

        World::getInstance()->getWorldFactory()->destroy(object);
      }
  }

  void PhysicsBody::load(PhysicsBody &object, lua_State *L, int index)
  {
    // Push another reference to the table on top of the stack (so we know
    // where it is, and this function can work for negative, positive and
    // pseudo indices
    lua_pushvalue(L, index);
    // stack now contains: -1 => table
    lua_pushnil(L);
    // stack now contains: -1 => nil; -2 => table
    while (lua_next(L, -2))
      {
        // stack now contains: -1 => value; -2 => key; -3 => table
        // copy the key so that lua_tostring does not modify the original
        lua_pushvalue(L, -2);
        // stack now contains: -1 => key; -2 => value; -3 => key; -4 => table
        const char *key = lua_tostring(L, -1);
        //            const char *value = lua_tostring(L, -2);
        if (lua_istable(L, -2))
          {
            PhysicsBody::load(object, L, -2);
          }
        else
          {
            if (lua_isnumber(L, index))
              {
                double number = lua_tonumber(L, index);
                printf("%s => %f\n", key, number);
              }
            else if (lua_isstring(L, index))
              {
                const char *v = lua_tostring(L, index);
                printf("%s => %s\n", key, v);
              }
            else if (lua_isboolean(L, index))
              {
                bool v = lua_toboolean(L, index);
                printf("%s => %d\n", key, v);
              }
            else if (lua_isuserdata(L, index))
              {
                //                    swig_lua_userdata *usr;
                //                    swig_type_info *type;
                //                    assert(lua_isuserdata(L,index));
                //                    usr=(swig_lua_userdata*)lua_touserdata(L,index);
                //                    /* get data */
                //                    type = usr->type;
                //                    njli::AbstractFactoryObject *object =
                //                    static_cast<njli::AbstractFactoryObject*>(usr->ptr);
                //                    printf("%s => %d:%s\n", key,
                //                    object->getType(),
                //                    object->getClassName());
              }
          }
        // pop value + copy of key, leaving original key
        lua_pop(L, 2);
        // stack now contains: -1 => key; -2 => table
      }
    // stack now contains: -1 => table (when lua_next returns 0 it pops the key
    // but does not push anything.)
    // Pop table
    lua_pop(L, 1);
    // Stack is now the same as it was on entry to this function
  }

  u32 PhysicsBody::type() { return JLI_OBJECT_TYPE_PhysicsBody; }

  void PhysicsBody::handleCollide(PhysicsBody *body, const btManifoldPoint &pt)
  {
    if (body->getPhysicsShape() && m_PhysicsShape)
      {
        char buffer[BUFFER_SIZE] = "__NJLINodeCollide";
        njli::World::getInstance()->getWorldLuaVirtualMachine()->execute(
            buffer, getParent(), body->getParent(), pt);
      }
  }

  void PhysicsBody::handleCollisionNear(PhysicsBody *body,
                                        const btDispatcherInfo &dispatchInfo)
  {
    if (m_PhysicsShape)
      {
        char buffer[BUFFER_SIZE] = "__NJLINodeNear";
        njli::World::getInstance()->getWorldLuaVirtualMachine()->execute(
            buffer, getParent(), body->getParent(), dispatchInfo);
      }
  }
  //    void PhysicsBody::addPhysicsWorld(PhysicsWorld *world)
  //    {
  //        SDL_assert(world != NULL);
  //
  //        if(getPhysicsWorld())
  //            removeChild(getPhysicsWorld());
  //
  //        m_PhysicsWorld = world;
  //
  //        addChild(m_PhysicsWorld);
  //    }
  //
  //    void PhysicsBody::removePhysicsWorld()
  //    {
  //        if(getPhysicsWorld())
  //            removeChild(getPhysicsWorld());
  //
  //        m_PhysicsWorld = NULL;
  //    }
  //
  //    PhysicsWorld *PhysicsBody::getPhysicsWorld()
  //    {
  //        s32 idx = getChildIndex(m_PhysicsWorld);
  //        if(idx != -1)
  //            return dynamic_cast<PhysicsWorld*>(getChild(idx));
  //        return NULL;
  //    }
  //
  //    const PhysicsWorld *PhysicsBody::getPhysicsWorld() const
  //    {
  //        s32 idx = getChildIndex(m_PhysicsWorld);
  //        if(idx != -1)
  //            return dynamic_cast<const PhysicsWorld*>(getChild(idx));
  //        return NULL;
  //    }

  void PhysicsBody::setPhysicsShape(PhysicsShape *shape)
  {
    SDL_assert(shape != NULL);

    if (getPhysicsShape())
      removeChild(getPhysicsShape());

    m_PhysicsShape = shape;

    addChild(m_PhysicsShape);

    //        Node *node = getParent();
    //        Geometry *geometry = node->getGeometry();
    //        if(geometry)
    //        {
    //            geometry->setTransform(node->getGeometryIndex(),
    //            node->getWorldTransform());
    //            PhysicsBody *physicsBody = node->getPhysicsBody();
    //            if(physicsBody)
    //                geometry->applyShape(node,
    //                physicsBody->getPhysicsShape());
    //        }
  }

  void PhysicsBody::removePhysicsShape()
  {
    if (getPhysicsShape())
      {
        removeChild(getPhysicsShape());
      }

    m_PhysicsShape = NULL;
  }

  PhysicsShape *PhysicsBody::getPhysicsShape()
  {
    s32 idx = getChildIndex(m_PhysicsShape);
    if (idx != -1)
      return dynamic_cast<PhysicsShape *>(getChild(idx));
    return NULL;
  }

  const PhysicsShape *PhysicsBody::getPhysicsShape() const
  {
    s32 idx = getChildIndex(m_PhysicsShape);
    if (idx != -1)
      {
        //            SDL_assert(dynamic_cast<const
        //            PhysicsShape*>(getChild(idx)));
        return dynamic_cast<const PhysicsShape *>(getChild(idx));
      }
    return NULL;
  }

  btTransform PhysicsBody::getWorldTransform() const
  {
    return getCollisionObject()->getWorldTransform();
  }

  void PhysicsBody::setFriction(f32 friction)
  {
    getCollisionObject()->setFriction(friction);
  }

  f32 PhysicsBody::getFriction() const
  {
    return getCollisionObject()->getFriction();
  }

  void PhysicsBody::setAngularFriction(f32 friction)
  {
    getCollisionObject()->setRollingFriction(friction);
  }

  f32 PhysicsBody::getAngularFriction() const
  {
    return getCollisionObject()->getRollingFriction();
  }

  void PhysicsBody::setRestitution(f32 restitution)
  {
    getCollisionObject()->setRestitution(restitution);
  }

  f32 PhysicsBody::getRestitution() const
  {
    return getCollisionObject()->getRestitution();
  }

  void PhysicsBody::setPhysicsType(njliPhysicsType type)
  {
    SDL_assert(type > JLI_PHYSICS_TYPE_NONE && type < JLI_PHYSICS_TYPE_MAX);

    switch (type)
      {
      case JLI_PHYSICS_TYPE_Dynamic:
        setDynamicPhysics();
        break;
      case JLI_PHYSICS_TYPE_Kinematic:
        setKinematicPhysics();
        break;
      case JLI_PHYSICS_TYPE_Static:
        setStaticPhysics();
        break;
      default:
        break;
      }

    if (PhysicsBody::getParent() && getPhysicsShape())
      {
        btTransform t = getWorldTransform();
        setTransform(t);
      }
  }

  njliPhysicsType PhysicsBody::getPhysicsType() const
  {
    if (isKinematicPhysics())
      return JLI_PHYSICS_TYPE_Kinematic;

    if (isStaticPhysics())
      return JLI_PHYSICS_TYPE_Static;

    if (isDynamicPhysics())
      return JLI_PHYSICS_TYPE_Dynamic;

    return JLI_PHYSICS_TYPE_NONE;
  }

  void PhysicsBody::setCollisionGroup(njliBitCategories categoryBitMask)
  {
    m_CategoryBitMask = categoryBitMask;
    if (PhysicsBody::getParent() && getPhysicsShape())
      {
        btTransform t = getWorldTransform();
        setTransform(t);
      }
  }

  njliBitCategories PhysicsBody::getCollisionGroup() const
  {
    return m_CategoryBitMask;
  }

  void PhysicsBody::setCollisionMask(njliBitCategories categoryBitMask)
  {
    m_CollisionBitMask = categoryBitMask;
    if (PhysicsBody::getParent() && getPhysicsShape())
      {
        btTransform t = getWorldTransform();
        setTransform(t);
      }
  }

  njliBitCategories PhysicsBody::getCollisionMask() const
  {
    return m_CollisionBitMask;
  }

  void PhysicsBody::setStaticPhysics()
  {
    m_CollisionFlags =
        On(m_CollisionFlags, btCollisionObject::CF_STATIC_OBJECT);
    m_CollisionFlags =
        Off(m_CollisionFlags, btCollisionObject::CF_KINEMATIC_OBJECT);
    m_ActivationState = ACTIVE_TAG;

    if (getCollisionObject())
      {
        getCollisionObject()->setCollisionFlags(m_CollisionFlags);
        getCollisionObject()->setActivationState(m_ActivationState);
      }

    if (PhysicsBody::getParent() && getPhysicsShape())
      {
        btTransform t = getWorldTransform();
        setTransform(t);
      }
  }
  void PhysicsBody::setKinematicPhysics()
  {
    m_CollisionFlags =
        On(m_CollisionFlags, btCollisionObject::CF_KINEMATIC_OBJECT);
    m_CollisionFlags =
        Off(m_CollisionFlags, btCollisionObject::CF_STATIC_OBJECT);
    m_ActivationState = DISABLE_DEACTIVATION;

    if (getCollisionObject())
      {
        getCollisionObject()->setCollisionFlags(m_CollisionFlags);
        getCollisionObject()->setActivationState(m_ActivationState);
      }

    if (PhysicsBody::getParent() && getPhysicsShape())
      {
        btTransform t = getWorldTransform();
        setTransform(t);
      }
  }
  void PhysicsBody::setDynamicPhysics()
  {
    m_CollisionFlags =
        Off(m_CollisionFlags, btCollisionObject::CF_KINEMATIC_OBJECT);
    m_CollisionFlags =
        Off(m_CollisionFlags, btCollisionObject::CF_STATIC_OBJECT);
    m_ActivationState = ACTIVE_TAG;

    if (getCollisionObject())
      {
        getCollisionObject()->setCollisionFlags(m_CollisionFlags);
        getCollisionObject()->setActivationState(m_ActivationState);
      }

    if (PhysicsBody::getParent() && getPhysicsShape())
      {
        btTransform t = getWorldTransform();
        setTransform(t);
      }
  }
  void PhysicsBody::enableContactResponse(bool enable)
  {
    if (enable)
      {
        m_CollisionFlags =
            Off(m_CollisionFlags, btCollisionObject::CF_NO_CONTACT_RESPONSE);
      }
    else
      {
        m_CollisionFlags =
            On(m_CollisionFlags, btCollisionObject::CF_NO_CONTACT_RESPONSE);
      }

    if (getCollisionObject())
      {
        getCollisionObject()->setCollisionFlags(m_CollisionFlags);
      }
  }
  void PhysicsBody::enableHandleCollideCallback(bool enable)
  {
    if (enable)
      {
        m_CollisionFlags = On(m_CollisionFlags,
                              btCollisionObject::CF_CUSTOM_MATERIAL_CALLBACK);
      }
    else
      {
        m_CollisionFlags = Off(m_CollisionFlags,
                               btCollisionObject::CF_CUSTOM_MATERIAL_CALLBACK);
      }

    if (getCollisionObject())
      {
        getCollisionObject()->setCollisionFlags(m_CollisionFlags);
      }
  }

  bool PhysicsBody::isStaticPhysics() const
  {
    return IsOn(m_CollisionFlags, btCollisionObject::CF_STATIC_OBJECT);
  }

  bool PhysicsBody::isKinematicPhysics() const
  {
    return IsOn(m_CollisionFlags, btCollisionObject::CF_KINEMATIC_OBJECT);
  }

  bool PhysicsBody::isDynamicPhysics() const
  {
    return (!(IsOn(m_CollisionFlags, btCollisionObject::CF_STATIC_OBJECT)) &&
            !(IsOn(m_CollisionFlags, btCollisionObject::CF_KINEMATIC_OBJECT)));
  }

  bool PhysicsBody::isResting() const { return m_IsResting; }

  void PhysicsBody::enableResting(bool enable) { m_RestingAllowed = enable; }

  s32 PhysicsBody::setPhysicsConstraint(PhysicsConstraint *constraint)
  {
    SDL_assert(NULL != constraint);

    Scene *scene = getParent()->getCurrentScene();
    SDL_assert(NULL != scene);

    if (getParent() && getParent()->getParentNode())
      {
        SDL_LogWarn(
            SDL_LOG_CATEGORY_TEST, "%s - %s->%s",
            "**ADDING A CONSTRAINT TO A CHILD NODE OTHER THAN THE ROOT**",
            getParent()->getName(), getParent()->getParentNode()->getName());
      }

    //        SDL_LogWarn(SDL_LOG_CATEGORY_TEST, "%s", (getParent() &&
    //        getParent()->getParentNode())?"**ADDING A CONSTRAINT TO A CHILD
    //        NODE OTHER THAN THE ROOT**\n":"");

    std::vector<PhysicsConstraint *>::const_iterator iter =
        std::find(m_PhysicsConstraintList.begin(),
                  m_PhysicsConstraintList.end(), constraint);

    if (iter == m_PhysicsConstraintList.end())
      {
        std::vector<PhysicsConstraint *>::iterator iter =
            std::find(m_PhysicsConstraintList.begin(),
                      m_PhysicsConstraintList.end(), constraint);

        if (iter != m_PhysicsConstraintList.end())
          removeChild(constraint);

        m_PhysicsConstraintList.push_back(constraint);

        addChild(constraint);
        //!!!TODO:
        //!!!TODO:            scene->addActivePhysicsConstraint(constraint);
      }

    return getPhysicsConstraintIndex(constraint);
  }

  bool PhysicsBody::removePhysicsConstraint(PhysicsConstraint *constraint)
  {
    SDL_assert(NULL != constraint);

    Scene *scene = getParent()->getCurrentScene();
    SDL_assert(NULL != scene);

    std::vector<PhysicsConstraint *>::iterator iter =
        std::find(m_PhysicsConstraintList.begin(),
                  m_PhysicsConstraintList.end(), constraint);

    if (iter != m_PhysicsConstraintList.end())
      {
        removeChild(*iter);
        //!!!TODO:            scene->removeActivePhysicsConstraint(*iter);

        m_PhysicsConstraintList.erase(iter);
        return true;
      }
    return false;
  }

  void PhysicsBody::removeAllPhysicsConstraints()
  {
    for (std::vector<PhysicsConstraint *>::iterator iter =
             m_PhysicsConstraintList.begin();
         iter != m_PhysicsConstraintList.end(); ++iter)
      {
        removeChild(*iter);
      }
    m_PhysicsConstraintList.clear();
  }

  void PhysicsBody::getPhysicsConstraints(
      std::vector<PhysicsConstraint *> &particleEmitters) const
  {
    for (std::vector<PhysicsConstraint *>::const_iterator iter =
             m_PhysicsConstraintList.begin();
         iter != m_PhysicsConstraintList.end(); ++iter)
      {
        if (getChildIndex(*iter) != -1)
          particleEmitters.push_back(*iter);
      }
  }

  s32 PhysicsBody::getPhysicsConstraintIndex(PhysicsConstraint *sound) const
  {
    std::vector<PhysicsConstraint *>::const_iterator iter = std::find(
        m_PhysicsConstraintList.begin(), m_PhysicsConstraintList.end(), sound);

    if (iter != m_PhysicsConstraintList.end())
      {
        return std::distance(m_PhysicsConstraintList.begin(), iter);
      }
    return -1;
  }

  PhysicsConstraint *PhysicsBody::getPhysicsConstraint(const u32 index)
  {
    if (index < m_PhysicsConstraintList.size())
      {
        s32 idx = getChildIndex(m_PhysicsConstraintList.at(index));
        if (idx != -1)
          return dynamic_cast<PhysicsConstraint *>(getChild(idx));
      }
    return NULL;
  }

  const PhysicsConstraint *
  PhysicsBody::getPhysicsConstraint(const u32 index) const
  {
    if (index < m_PhysicsConstraintList.size())
      {
        s32 idx = getChildIndex(m_PhysicsConstraintList.at(index));
        if (idx != -1)
          return dynamic_cast<const PhysicsConstraint *>(getChild(idx));
      }
    return NULL;
  }

  bool PhysicsBody::hasPhysicsConstraint() const
  {
    unsigned long constraintListSize = m_PhysicsConstraintList.size();
    const btRigidBody *rigidBody =
        dynamic_cast<const btRigidBody *>(getCollisionObject());

    SDL_assert(rigidBody->getNumConstraintRefs() == constraintListSize);

    //        SDL_LogVerbose(SDL_LOG_CATEGORY_TEST, "hasPhysicsConstraint: %d",
    //        s);
    return constraintListSize > 0;
  }

  //    void PhysicsBody::enablePropertyChange(bool enable)
  //    {
  //        m_PropertyChanged = enable;
  //    }
  //
  //    bool PhysicsBody::isPropertyChanged()const
  //    {
  //        if(getPhysicsShape())
  //            return m_PropertyChanged ||
  //            getPhysicsShape()->isPropertyChanged();
  //        return m_PropertyChanged;
  //    }

  Node *PhysicsBody::getParent()
  {
    return dynamic_cast<Node *>(AbstractDecorator::getParent());
  }

  const Node *PhysicsBody::getParent() const
  {
    return dynamic_cast<const Node *>(AbstractDecorator::getParent());
  }

  s32 PhysicsBody::getCollisionFlags() const { return m_CollisionFlags; }

  s32 PhysicsBody::getActivationState() const { return m_ActivationState; }
}
